### PR TITLE
Report result by player and referee

### PR DESCRIPTION
**Pull request by CioleaFlorin**

### CODE REVIEW

File: essentialprogramming-api/src/main/java/com/api/controller/MatchController.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+public class MatchController {
+    private final MatchService matchService;
+
+    @Autowired
+    public MatchController(MatchService matchService) {
+        this.matchService = matchService;
+    }
+
+    @POST
+    @Path("/player-report-match/{userKey}/{matchKey}/{result}")
```
Comment (gabrielhorgos) [inline essentialprogramming-api/src/main/java/com/api/controller/MatchController.java:38]: player-report should suffice. 
I think we should use path params when specifying resource and filtering /updatable parameters could be query params.
/match/{matchKey/report?playerKey=val&result=val

File: essentialprogramming-api/src/main/java/com/api/service/MatchService.java [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+@Service
+public class MatchService {
+
+    private final PlayerRepository playerRepository;
+    private final MatchRepository matchRepository;
+    private final MatchResultRepository matchResultRepository;
+    private final TournamentUserRepository tournamentUserRepository;
+
+
+    @Autowired
```
Comment (gabrielhorgos) [inline essentialprogramming-api/src/main/java/com/api/service/MatchService.java:29]: replace with lombok annotation and get rid of constructor

File: essentialprogramming-api/src/main/java/com/api/service/MatchService.java [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+        if (bothResultsPresentAndEqual(matchResult)) {
+            matchResult.setResult(result);
+            addPointToPlayers(matchResult.getFirstPlayer().getId(), matchResult.getSecondPlayer().getId(), result, currentMatch.getTournament().getId());
+            currentMatch.setState(GameState.ENDED);
+
+        } else {
+            //TODO: notify referee
+        }
+
+        matchResultRepository.save(matchResult);
```
Comment (gabrielhorgos) [inline essentialprogramming-api/src/main/java/com/api/service/MatchService.java:64]: is there no cascade or why do we need to call save on both ?

File: essentialprogramming-api/src/main/java/com/api/service/MatchService.java [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+            return Result.FIRST;
+        } else if (resultString.equals((Result.SECOND.getValue()))) {
+            return Result.SECOND;
+        } else if (resultString.equals(Result.DRAW.getValue())) {
+            return Result.DRAW;
+        } else
+            throw new HttpClientErrorException(HttpStatus.BAD_REQUEST, "Invalid result!");
+    }
+
+    private void addPointToPlayers(int firstPlayerId, int secondPlayerId, Result result, int tournamentId) {
```
Comment (gabrielhorgos) [inline essentialprogramming-api/src/main/java/com/api/service/MatchService.java:83]: just call it applyResults ?

File: essentialprogramming-api/src/main/java/com/api/service/MatchService.java [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+
+        Player firstPlayer = playerRepository.findById(firstPlayerId).orElseThrow(() -> new HttpClientErrorException(HttpStatus.NOT_FOUND, "Player not found!"));
+        Player secondPlayer = playerRepository.findById(secondPlayerId).orElseThrow(() -> new HttpClientErrorException(HttpStatus.NOT_FOUND, "Player not found!"));
+
+        TournamentUser tournamentUserFirstPlayer = tournamentUserRepository.findById(new TournamentUserKey(tournamentId, firstPlayerId))
+                .orElseThrow(() -> new HttpClientErrorException(HttpStatus.NOT_FOUND, "This player is not participating in the tournament"));
+        TournamentUser tournamentUserSecondPlayer = tournamentUserRepository.findById(new TournamentUserKey(tournamentId, secondPlayerId))
+                .orElseThrow(() -> new HttpClientErrorException(HttpStatus.NOT_FOUND, "This player is not participating in the tournament"));
+
+        if (result == Result.FIRST) {
```
Comment (gabrielhorgos) [inline essentialprogramming-api/src/main/java/com/api/service/MatchService.java:93]: use instead Result.FIRST.equals(result)
that way you can avoid nullpointer ex (NPE)

Comment (gabrielhorgos) [reply essentialprogramming-api/src/main/java/com/api/service/MatchService.java:93]: on a second thought you should create a switch here

Comment (gabrielhorgos) [reply essentialprogramming-api/src/main/java/com/api/service/MatchService.java:93]: save only user that has changed. Also it's .save required isn't it getting commited on transaction end ?

File: essentialprogramming-api/src/main/java/com/api/service/MatchService.java [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+            firstPlayer.setScore(firstPlayer.getScore() + 0.5);
+            secondPlayer.setScore(secondPlayer.getScore() + 0.5);
+            tournamentUserFirstPlayer.setScore(tournamentUserFirstPlayer.getScore() + 0.5);
+            tournamentUserSecondPlayer.setScore(tournamentUserSecondPlayer.getScore() + 0.5);
+        }
+        tournamentUserRepository.save(tournamentUserFirstPlayer);
+        tournamentUserRepository.save(tournamentUserSecondPlayer);
+    }
+
+    private boolean bothResultsPresentAndEqual(MatchResult matchResult) {
```
Comment (gabrielhorgos) [inline essentialprogramming-api/src/main/java/com/api/service/MatchService.java:109]: isADraw ?

Comment (CioleaFlorin) [reply essentialprogramming-api/src/main/java/com/api/service/MatchService.java:109]: it is a check to see if both players reported the match result and if they put the same result (all 3 options, not just a draw)

**Pull request merged**